src/test/java/com/google/devtools/build/lib/starlark/StarlarkRuleImplementationFunctionsTest.java

```diff
@@ -3480,6 +3480,178 @@ public void starlarkCustomCommandLineKeyComputation_inconsequentialChangeToStarl
+    setRuleContext(createRuleContext("//foo:foo"));
+
+    // Verify that a depset with elements of type Label and a list that starts with a Label and has
+    // elements with identical string representation to those of the depset have different keys.
+    RepositoryMapping mainRepoMapping =
+        RepositoryMapping.create(
+            ImmutableMap.of(
+                "apparent1",
+                RepositoryName.createUnvalidated("canonical1~"),
+                "apparent2",
+                RepositoryName.createUnvalidated("canonical2~")),
+            RepositoryName.MAIN);
+    CommandLine commandLine1 =
+        getCommandLine(
+            mainRepoMapping,
+            """
+            args = ruleContext.actions.args()
+            args.add_all(depset([Label("@@canonical1~//foo:bar"), Label("@@canonical2~//foo:bar")]))
+            """);
+    CommandLine commandLine2 =
+        getCommandLine(
+            mainRepoMapping,
+            """
+            args = ruleContext.actions.args()
+            args.add_all([Label("@@canonical1~//foo:bar"), str(Label("@@canonical2~//foo:bar"))])
+            """);
+
+    assertThat(getArguments(commandLine1, PathMapper.NOOP))
+        .isNotEqualTo(getArguments(commandLine2, PathMapper.NOOP));
+    System.err.println("starting hashes");
+    assertThat(getDigest(commandLine1)).isEqualTo(getDigest(commandLine2));
+  }
+
+  @Test
+  public void starlarkCustomCommandLineKeyComputation_artifactVsPathStringInAdd() throws Exception {
+    setRuleContext(createRuleContext("//foo:foo"));
+
+    CommandLine commandLine1 =
+        getCommandLine(
+            """
+                    file = ruleContext.actions.declare_file('file')
+                    args = ruleContext.actions.args()
+                    args.add(file)
+                    """);
+    CommandLine commandLine2 =
+        getCommandLine(
+            """
+                    file = ruleContext.actions.declare_file('file')
+                    args = ruleContext.actions.args()
+                    args.add(file.path)
+                    """);
+
+    assertThat(getArguments(commandLine1, PathMapper.NOOP))
+        .isEqualTo(getArguments(commandLine2, PathMapper.NOOP));
+    assertThat(getArguments(commandLine1, NON_TRIVIAL_PATH_MAPPER))
+        .isNotEqualTo(getArguments(commandLine2, NON_TRIVIAL_PATH_MAPPER));
+    assertThat(getDigest(commandLine1)).isNotEqualTo(getDigest(commandLine2));
+  }
+
+  @Test
+  public void starlarkCustomCommandLineKeyComputation_artifactVsPathStringInAddFormatted()
+      throws Exception {
+    setRuleContext(createRuleContext("//foo:foo"));
+
+    CommandLine commandLine1 =
+        getCommandLine(
+            """
+                    file = ruleContext.actions.declare_file('file')
+                    args = ruleContext.actions.args()
+                    args.add(file, format = '--%s')
+                    """);
+    CommandLine commandLine2 =
+        getCommandLine(
+            """
```

Comment:
@cushon The IntelliJ google-java-format plugin doesn't seem to support text blocks that well yet. What is the recommended style here? I also couldn't find any reference to text blocks in the Google Java Style Guide.

Reply:
I think a version was just published that does support them, and the CLI should already: https://github.com/google/google-java-format/issues/1024#issuecomment-2056916792

Anyway the current approach is e.g.

```java
  CommandLine commandLine1 =
      getCommandLine(
          """
          file = ruleContext.actions.declare_file('file')
          args = ruleContext.actions.args()
          args.add(file, format = '--%s')
          """);
```

Reply:
Thanks!